Python Assignment 1

To Analyse and Classify the Data Set Credit Card based on multiple dataframe columns and various Classification models.

[Link for the Accuracies of models I tried training for the dataset](https://docs.google.com/spreadsheets/d/1s2lJFry7ITm2CaDpH20vz4f2WeHKmERvLuEaEOUPlRA/edit?usp=sharing)

Step 1: Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt

Step 2: Importing the data and printing the dataset attributes

In [2]:
dataset = pd.read_csv('Credit_Card.csv')
dataset.head()

Unnamed: 0         X1   X2         X3        X4   X5     X6     X7     X8  \
0         ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3   
1          1      20000    2          2         1   24      2      2     -1   
2          2     120000    2          2         2   26     -1      2      0   
3          3      90000    2          2         2   34      0      0      0   
4          4      50000    2          2         1   37      0      0      0   

      X9  ...        X15        X16        X17       X18       X19       X20  \
0  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3   
1     -1  ...          0          0          0         0       689         0   
2      0  ...       3272       3455       3261         0      1000      1000   
3      0  ...      14331      14948      15549      1518      1500      1000   
4      0  ...      28314      28959      29547      2000      2019      1200   

        X21       X22       X23                           Y  
0  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1         0         0         0                           1  
2      1000         0      2000                           1  
3      1000      1000      5000                           0  
4      1100      1069      1000                           0  

[5 rows x 25 columns]

We can see that There is Row 0 which has column titles.

Step 3: Data Analysis

Step 3a: We need to either remove the level X1, X2, X3 or the column names so that we have indexed the attributes properly. I will go with keeping Column titles for analysis. And for the actual Classification we don't really access them by their name in classification.

In [3]:
dataset.columns = dataset.iloc[0]
dataset = dataset[1:]
dataset.head()

0 ID LIMIT_BAL SEX EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3 PAY_4  ...  \
1  1     20000   2         2        1  24     2     2    -1    -1  ...   
2  2    120000   2         2        2  26    -1     2     0     0  ...   
3  3     90000   2         2        2  34     0     0     0     0  ...   
4  4     50000   2         2        1  37     0     0     0     0  ...   
5  5     50000   1         2        1  57    -1     0    -1     0  ...   

0 BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 PAY_AMT3 PAY_AMT4 PAY_AMT5  \
1         0         0         0        0      689        0        0        0   
2      3272      3455      3261        0     1000     1000     1000        0   
3     14331     14948     15549     1518     1500     1000     1000     1000   
4     28314     28959     29547     2000     2019     1200     1100     1069   
5     20940     19146     19131     2000    36681    10000     9000      689   

0 PAY_AMT6 default payment next month  
1        0                          1  
2     2000                          1  
3     5000                          0  
4     1000                          0  
5      679                          0  

[5 rows x 25 columns]

Step 3b: Data Shapes and dtypes


In [4]:
dataset.shape

(30000, 25)

we can see there are 30k rows and 25 columns

Set ID as Index

In [5]:
dataset = dataset.set_index('ID')
dataset.head()

0  LIMIT_BAL SEX EDUCATION MARRIAGE AGE PAY_0 PAY_2 PAY_3 PAY_4 PAY_5  ...  \
ID                                                                     ...   
1      20000   2         2        1  24     2     2    -1    -1    -2  ...   
2     120000   2         2        2  26    -1     2     0     0     0  ...   
3      90000   2         2        2  34     0     0     0     0     0  ...   
4      50000   2         2        1  37     0     0     0     0     0  ...   
5      50000   1         2        1  57    -1     0    -1     0     0  ...   

0  BILL_AMT4 BILL_AMT5 BILL_AMT6 PAY_AMT1 PAY_AMT2 PAY_AMT3 PAY_AMT4 PAY_AMT5  \
ID                                                                              
1          0         0         0        0      689        0        0        0   
2       3272      3455      3261        0     1000     1000     1000        0   
3      14331     14948     15549     1518     1500     1000     1000     1000   
4      28314     28959     29547     2000     2019     1200     1100     1069   
5      20940     19146     19131     2000    36681    10000     9000      689   

0  PAY_AMT6 default payment next month  
ID                                      
1         0                          1  
2      2000                          1  
3      5000                          0  
4      1000                          0  
5       679                          0  

[5 rows x 24 columns]

In [6]:

dataset.shape

(30000, 24)

In [7]:
dataset.dtypes

0
LIMIT_BAL                     object
SEX                           object
EDUCATION                     object
MARRIAGE                      object
AGE                           object
PAY_0                         object
PAY_2                         object
PAY_3                         object
PAY_4                         object
PAY_5                         object
PAY_6                         object
BILL_AMT1                     object
BILL_AMT2                     object
BILL_AMT3                     object
BILL_AMT4                     object
BILL_AMT5                     object
BILL_AMT6                     object
PAY_AMT1                      object
PAY_AMT2                      object
PAY_AMT3                      object
PAY_AMT4                      object
PAY_AMT5                      object
PAY_AMT6                      object
default payment next month    object
dtype: object

In [8]:
dataset = dataset.astype(int)

In [9]:
dataset.dtypes

0
LIMIT_BAL                     int64
SEX                           int64
EDUCATION                     int64
MARRIAGE                      int64
AGE                           int64
PAY_0                         int64
PAY_2                         int64
PAY_3                         int64
PAY_4                         int64
PAY_5                         int64
PAY_6                         int64
BILL_AMT1                     int64
BILL_AMT2                     int64
BILL_AMT3                     int64
BILL_AMT4                     int64
BILL_AMT5                     int64
BILL_AMT6                     int64
PAY_AMT1                      int64
PAY_AMT2                      int64
PAY_AMT3                      int64
PAY_AMT4                      int64
PAY_AMT5                      int64
PAY_AMT6                      int64
default payment next month    int64
dtype: object

I checked if the dataset is balanced or imbalanced.
Using following:

In [10]:
dataset['IsDef'] = dataset['default payment next month']
dataset.drop('default payment next month',axis = 1)
dataset['IsDef'].value_counts()


0    23364
1     6636
Name: IsDef, dtype: int64

Step 3c: Filter the required data(edit the columns)
Moving some columns to fit them in the Feature Scaling



We will need to check the range of each column

In [11]:
print(dataset.min())
print(dataset.max())

0
LIMIT_BAL                      10000
SEX                                1
EDUCATION                          0
MARRIAGE                           0
AGE                               21
PAY_0                             -2
PAY_2                             -2
PAY_3                             -2
PAY_4                             -2
PAY_5                             -2
PAY_6                             -2
BILL_AMT1                    -165580
BILL_AMT2                     -69777
BILL_AMT3                    -157264
BILL_AMT4                    -170000
BILL_AMT5                     -81334
BILL_AMT6                    -339603
PAY_AMT1                           0
PAY_AMT2                           0
PAY_AMT3                           0
PAY_AMT4                           0
PAY_AMT5                           0
PAY_AMT6                           0
default payment next month         0
IsDef                              0
dtype: int64
0
LIMIT_BAL                     1000000
SEX                 

Step 3d: I rearranged the columns according to the variables I want to consider.

For the optimum average accuracy I have taken
All payment values into consideration.
The explained average accuracy which can be seen in given Sheet in the first Text only.

In [12]:
data = dataset.iloc[:,[0,1,2,3,4,5,6,7,8,9,10,23]]
#all : ,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22
#first_all : ,5,11,17
#last_all : ,10,16,22
#first_last : ,5,10,11,16,17,22
#pay : ,5,6,7,8,9,10
#bal: ,11,12,13,14,15,16
#bill : ,17,18,19,20,21,22
#no_sex : remove 1
#no_limit_bal : remove 0
#no_edu : remove 2
#no_age : remove 3
#no_marrige : remove 4
data.head()

0   LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
ID                                                                         
1       20000    2          2         1   24      2      2     -1     -1   
2      120000    2          2         2   26     -1      2      0      0   
3       90000    2          2         2   34      0      0      0      0   
4       50000    2          2         1   37      0      0      0      0   
5       50000    1          2         1   57     -1      0     -1      0   

0   PAY_5  PAY_6  default payment next month  
ID                                            
1      -2     -2                           1  
2       0      2                           1  
3       0      0                           0  
4       0      0                           0  
5       0      0                           0

Step 4: Data Modelling, Now we have our rearranged Data. We can use it to model further.

Step 4a: We need to divide Data in X variable and y variable. Y variable will be the Default decision.

In [13]:
X = data.iloc[:,:-1].values
y = data.iloc[:,-1].values

In [14]:
print(X)

[[ 20000      2      2 ...     -1     -2     -2]
 [120000      2      2 ...      0      0      2]
 [ 90000      2      2 ...      0      0      0]
 ...
 [ 30000      1      2 ...     -1      0      0]
 [ 80000      1      3 ...      0      0     -1]
 [ 50000      1      2 ...      0      0      0]]


In [15]:
print(y)

[1 1 0 ... 1 1 1]


Step 4b: Divide the data in test set and train set with ration being 20% for Test and rest for the Train

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

Print All Datas divided

In [17]:
print(X_train)
len(X_train)

[[ 20000      2      2 ...      0      0      0]
 [ 50000      2      2 ...     -1     -1     -1]
 [180000      1      1 ...     -2     -2     -2]
 ...
 [ 50000      1      2 ...      0      0      0]
 [ 80000      1      1 ...     -1     -1     -1]
 [300000      1      2 ...     -1      0      0]]


24000

In [18]:
print(X_test)
len(X_test)

[[310000      1      3 ...      0      0      0]
 [ 10000      2      3 ...     -1      2      2]
 [ 50000      1      2 ...      0     -1     -1]
 ...
 [ 50000      1      2 ...      0      0      0]
 [ 20000      2      2 ...      0     -2     -2]
 [240000      2      2 ...     -1     -1     -1]]


6000

In [19]:
print(y_train)

[1 0 0 ... 0 0 0]


In [20]:
print(y_test)

[0 1 0 ... 0 0 0]


Step 4c: Adding a Feature scaling to certain datatypes so they do fit in the range. I already rearranged the data according to the data limits.

In [21]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

Check the Scaled Data

In [22]:
print(X_train)

[[-1.13724     0.8092862   0.18411875 ...  0.18684613  0.23307076
   0.25318848]
 [-0.90537456  0.8092862   0.18411875 ... -0.66633253 -0.64768791
  -0.61687502]
 [ 0.0993757  -1.23565681 -1.08130221 ... -1.51951119 -1.52844659
  -1.48693851]
 ...
 [-0.90537456 -1.23565681  0.18411875 ...  0.18684613  0.23307076
   0.25318848]
 [-0.67350911 -1.23565681 -1.08130221 ... -0.66633253 -0.64768791
  -0.61687502]
 [ 1.02683748 -1.23565681  0.18411875 ... -0.66633253  0.23307076
   0.25318848]]


Step 5: Adding the data to multiple classification models

Step 5a: Logistic Regression Model

In [23]:
from sklearn.linear_model import LogisticRegression
classifierlog = LogisticRegression(random_state = 0)
classifierlog.fit(X_train, y_train)

LogisticRegression(random_state=0)

Predict The data output for Test Data

In [24]:
y_pred_log = classifierlog.predict(X_test)
print(y_pred_log.reshape(len(y_pred_log),1))

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluate using the confusion matrix

In [25]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_log)
print(cm)
accuracy_score(y_test, y_pred_log)

[[4550  113]
 [1042  295]]


0.8075

The result in accuracy is: **80.75%**

Step 5b: K Nearest Neighbours Classification Model

In [26]:
from sklearn.neighbors import KNeighborsClassifier
classifierknn = KNeighborsClassifier(n_neighbors = 10, metric = 'minkowski', p = 2)
classifierknn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=10)

Predict The data or Test model

In [27]:
y_pred_knn = classifierknn.predict(X_test)
print(y_pred_knn.reshape(len(y_pred_knn),1))

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluate the Confusion Matrix and Accuracy

In [28]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_knn)
print(cm)
accuracy_score(y_test, y_pred_knn)

[[4442  221]
 [ 914  423]]


0.8108333333333333

The resuting accuracy is: **81.08%**

Step 5c: SVM

In [29]:
from sklearn.svm import SVC
classifiersvm = SVC(kernel = 'linear', random_state = 0)
classifiersvm.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

Predict the data for Test Set

In [30]:
y_pred_svm = classifiersvm.predict(X_test)
print(y_pred_svm.reshape(len(y_pred_svm),1))

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluate the confusion Matrix and Accuracy

In [31]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_svm)
print(cm)
accuracy_score(y_test, y_pred_svm)

[[4543  120]
 [1039  298]]


0.8068333333333333

The resulting accuracy is: **80.68%**

Step 5d: Kernel SVM

In [32]:
from sklearn.svm import SVC
classifierksvm = SVC(kernel = 'rbf', random_state = 0)
classifierksvm.fit(X_train, y_train)

SVC(random_state=0)

Predicting The Test Output

In [33]:
y_pred_ksvm = classifierksvm.predict(X_test)
print(y_pred_ksvm.reshape(len(y_pred_ksvm),1))

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluating the confusion Matrix and Accuracy of the data.

In [34]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_ksvm)
print(cm)
accuracy_score(y_test, y_pred_ksvm)

[[4460  203]
 [ 881  456]]


0.8193333333333334

The resulting Accuracy is: **81.93%**



Step 5e: Naive Bayes Classification

In [35]:
from sklearn.naive_bayes import GaussianNB
classifiernaive = GaussianNB()
classifiernaive.fit(X_train, y_train)

GaussianNB()

Predicting The Test output

In [36]:
y_pred_naive = classifiernaive.predict(X_test)
print(y_pred_naive.reshape(len(y_pred_naive),1))

[[0]
 [1]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluate the Confusion Matrix and Accuracy

In [37]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_naive)
print(cm)
accuracy_score(y_test, y_pred_naive)

[[4283  380]
 [ 798  539]]


0.8036666666666666

The resulting Accuracy is: **80.36%**

Step 5f: Decsion Tree Classification

In [38]:
from sklearn.tree import DecisionTreeClassifier
classifierdtc = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifierdtc.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

Predicting the Test Output

In [39]:
y_pred_dtc = classifierdtc.predict(X_test)
print(y_pred_dtc.reshape(len(y_pred_dtc),1))

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluation of Confusion Matrix and Accuracy:

In [40]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_dtc)
print(cm)
accuracy_score(y_test, y_pred_dtc)

[[3938  725]
 [ 836  501]]


0.7398333333333333

The Accuracy is: **73.98%**

Step 5g: Random Forest Classifier

In [41]:
from sklearn.ensemble import RandomForestClassifier
classifierRF = RandomForestClassifier(n_estimators = 20, criterion = 'entropy', random_state = 0)
classifierRF.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=20, random_state=0)

Predict The Test Data

In [42]:
y_pred_rf = classifierRF.predict(X_test)
print(y_pred_rf.reshape(len(y_pred_rf),1))

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Evaluate the Confusion Matrix and Accuracy





In [43]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred_rf)
print(cm)
accuracy_score(y_test, y_pred_rf)

[[4275  388]
 [ 886  451]]


0.7876666666666666

The resulting Accuracy is: **78.76%**

However if we change the number of estimators from 20 to let's say, 10. Which basically changes the number of decision trees we have made. Then the accuracy might change. It is generalised that the higher n_esitmators means better coverage and more accurate predictions but it also has higher constraints over computations.

Section 6: Evaluation of All Accuracies and result

All results were found to lie in Accuracy range of 71% to 82%(As in range 70-80%) when all the variables are included.

According to the results found, The Kernel SVM Classifier shows the best results in terms of accuracy at 81.93%.

Why It Might be that way?

A: According to me, the KSVM does not bias Outliers like some other Classifiers. It also does not have an issue of overfitting. Hence it evaluates each and every dataset attribute closely and accurately.

However it takes a huge computational time and space.

Also, some variable selection shows slight improvement over others selections as well.